File: api/routes/apply.ts
/**
 * Apply route handler (public).
 * GET  /api/apply/job/:jobId
 * POST /api/apply/submit
 * Extracted verbatim from api/[...path].ts — lines 5476-5722.
 */
import type { VercelRequest, VercelResponse } from '@vercel/node';
import { getSupabaseAdmin } from '../_lib/supabase';
import { ok, badRequest, notFound, methodNotAllowed } from '../_lib/helpers';
import { generateJSON } from '../_lib/openai';
import { extractResumeText } from '../_lib/resume-parser';
import { sendApplicationReceived } from '../_lib/offer-utils';
import { v4 as uuidv4 } from 'uuid';
import Busboy from 'busboy';

export default async function handleApply(req: VercelRequest, res: VercelResponse, segments: string[]) {
  const supabase = getSupabaseAdmin();

  // GET /api/apply/job/:jobId - Public job details
  if (segments.length === 3 && segments[1] === 'job') {
    if (req.method !== 'GET') return methodNotAllowed(res);

    const jobId = segments[2];
    const { data: job, error } = await supabase
      .from('job_descriptions')
      .select('id, title, role, level, description, must_have_skills, good_to_have_skills, min_experience_years, is_active')
      .eq('id', jobId)
      .eq('is_active', true)
      .single();

    if (error || !job) return notFound(res, 'Job not found or no longer accepting applications');
    return ok(res, job);
  }

  // POST /api/apply/submit - Submit job application (multipart form)
  if (segments.length === 2 && segments[1] === 'submit') {
    if (req.method !== 'POST') return methodNotAllowed(res);

    // Parse multipart form data
    const fields: Record<string, string> = {};
    let resumeFile: { filename: string; mimeType: string; buffer: Buffer } | null = null;

    await new Promise<void>((resolve, reject) => {
      const ct = req.headers['content-type'] || '';
      if (!ct.includes('multipart/form-data')) {
        // JSON body fallback
        const body = req.body || {};
        Object.entries(body).forEach(([k, v]) => { fields[k] = String(v); });
        resolve();
        return;
      }

      const bb = Busboy({ headers: req.headers as any });
      bb.on('field', (name: string, val: string) => { fields[name] = val; });
      bb.on('file', (name: string, file: any, info: any) => {
        if (name === 'resume') {
          const chunks: Buffer[] = [];
          file.on('data', (d: Buffer) => chunks.push(Buffer.from(d)));
          file.on('end', () => {
            resumeFile = {
              filename: info.filename || 'resume.pdf',
              mimeType: info.mimeType || 'application/octet-stream',
              buffer: Buffer.concat(chunks),
            };
          });
        } else {
          file.resume();
        }
      });
      bb.on('finish', () => resolve());
      bb.on('error', (err: Error) => reject(err));
      req.pipe(bb);
    });

    const jobId = fields.job_id;
    const fullName = fields.full_name;
    const email = fields.email;
    const phone = fields.phone || null;
    const portfolioUrl = fields.portfolio_url || null;
    const githubUrl = fields.github_url || null;
    const location = fields.location || null;
    const vendorName = fields.vendorName || null;
    const mainSkillset = fields.mainSkillset || null;
    const consentGiven = fields.consent_given === 'true';

    if (!jobId || !fullName || !email) {
      return badRequest(res, 'job_id, full_name, and email are required');
    }

    if (!consentGiven) {
      return badRequest(res, 'Consent is required to submit application');
    }

    // Verify job exists and is active
    const { data: job } = await supabase
      .from('job_descriptions')
      .select('*')
      .eq('id', jobId)
      .eq('is_active', true)
      .single();

    if (!job) return notFound(res, 'Job not found or no longer accepting applications');

    // Check if candidate already exists
    const { data: existingCandidates } = await supabase
      .from('candidates')
      .select('id')
      .eq('email', email)
      .limit(1);

    let candidateId: string;

    if (existingCandidates && existingCandidates.length > 0) {
      candidateId = existingCandidates[0].id;

      // Check if already applied to this job
      const { data: existingApp } = await supabase
        .from('job_applications')
        .select('id')
        .eq('candidate_id', candidateId)
        .eq('job_id', jobId)
        .limit(1);

      if (existingApp && existingApp.length > 0) {
        return badRequest(res, 'You have already applied to this job');
      }

      // Update candidate info
      await supabase
        .from('candidates')
        .update({
          full_name: fullName,
          phone,
          portfolio_url: portfolioUrl,
          github_url: githubUrl,
          location,
          vendorName,
          mainSkillset,
          consent_given: consentGiven,
          consent_timestamp: new Date().toISOString(),
          updated_at: new Date().toISOString(),
        })
        .eq('id', candidateId);
    } else {
      // Create new candidate
      const candidateData: Record<string, any> = {
        full_name: fullName,
        email,
        phone,
        portfolio_url: portfolioUrl,
        github_url: githubUrl,
        location,
        vendorName,
        mainSkillset,
        consent_given: consentGiven,
        consent_timestamp: new Date().toISOString(),
      };

      const { data: newCandidate, error: createErr } = await supabase
        .from('candidates')
        .insert(candidateData)
        .select()
        .single();

      if (createErr || !newCandidate) {
        return res.status(500).json({ error: createErr?.message || 'Failed to create candidate' });
      }
      candidateId = newCandidate.id;
    }

    // Handle resume upload
    let resumeParsedData: any = null;
    let resumeText = '';
    if (resumeFile) {
      try {
        const uploadedResume = resumeFile as { filename: string; mimeType: string; buffer: Buffer };
        const rawText = await extractResumeText(uploadedResume.buffer, uploadedResume.filename);
        resumeText = String(rawText)
          .replace(/\x00/g, '')
          .replace(/[\x00-\x08\x0B\x0C\x0E-\x1F]/g, '')
          .slice(0, 50000);

        // Parse with AI
        try {
          const prompt = \`You are an expert resume parser.\n\nParse the following resume and return ONLY valid JSON in this exact format:\n{\n  "skills": ["skill1"],\n  "experience": [{"title":"","company":"","duration":"","description":""}],\n  "education": [{"degree":"","institution":"","year":""}],\n  "summary": "",\n  "total_experience_years": 0,\n  "certifications": ["cert1"]\n}\n\nRESUME TEXT:\n\${resumeText.slice(0, 8000)}\`;
          resumeParsedData = await generateJSON<any>(prompt);
        } catch (err: any) {
          console.error('Resume AI parsing failed during application (will save raw text anyway):', err?.message);
          resumeParsedData = null;
        }

        // Update candidate with resume data
        await supabase
          .from('candidates')
          .update({
            resume_text: resumeText,
            resume_parsed_data: resumeParsedData,
            updated_at: new Date().toISOString(),
          })
          .eq('id', candidateId);
      } catch (e: any) {
        console.error('Resume parsing error during apply:', e?.message);
      }
    }

    // Create job application
    const applicationId = uuidv4();
    const { error: appError } = await supabase
      .from('job_applications')
      .insert({
        id: applicationId,
        candidate_id: candidateId,
        job_id: jobId,
        status: 'applied',
        applied_at: new Date().toISOString(),
      });

    if (appError) {
      return res.status(500).json({ error: appError.message });
    }

    // Run ATS screening if resume was parsed
    if (resumeParsedData) {
      try {
        const prompt = \`Analyze this candidate's resume against the job requirements and provide ATS screening scores.\n\nJob: \${job.title} (\${job.role}, \${job.level})\nRequired Skills: \${(job.must_have_skills || []).join(', ')}\nNice-to-have Skills: \${(job.good_to_have_skills || []).join(', ')}\nMin Experience: \${job.min_experience_years} years\n\nCandidate Resume JSON:\n\${JSON.stringify(resumeParsedData)}\n\nReturn JSON:\n{\n  "overall_score": 0-100,\n  "skill_relevance_score": 0-100,\n  "experience_score": 0-100,\n  "education_score": 0-100,\n  "credibility_score": 0-100,\n  "shortlisted": true/false,\n  "shortlist_reason": "...",\n  "reason_codes": [{"code":"SKILL_MATCH","type":"positive","description":"...","impact":10}]\n}\`;
        const screeningResult = await generateJSON<any>(prompt);

        await supabase.from('ats_screenings').insert({
          candidate_id: candidateId,
          job_id: jobId,
          overall_score: screeningResult.overall_score,
          skill_relevance_score: screeningResult.skill_relevance_score ?? null,
          experience_score: screeningResult.experience_score ?? null,
          education_score: screeningResult.education_score ?? null,
          credibility_score: screeningResult.credibility_score ?? null,
          shortlisted: !!screeningResult.shortlisted,
          shortlist_reason: screeningResult.shortlist_reason ?? null,
          reason_codes: screeningResult.reason_codes ?? [],
          screened_at: new Date().toISOString(),
        });
      } catch (e: any) {
        console.error('ATS screening error during apply:', e?.message);
      }
    }

    // Send confirmation email
    try {
      await sendApplicationReceived(email, fullName, job.title);
    } catch (e: any) {
      console.error('Failed to send confirmation email:', e?.message);
    }

    return ok(res, {
      id: applicationId,
      job_id: jobId,
      candidate_id: candidateId,
      status: 'applied',
      message: \`Application submitted successfully for \${job.title}. Check your email for confirmation.\`,
    }, 201);
  }

  return notFound(res);
}

File: backend/app/api/v2/endpoints/apply.py
from __future__ import annotations

import logging
from datetime import datetime, timezone
from typing import Any, Dict, Optional

from fastapi import APIRouter, File, Form, UploadFile, HTTPException

from app.services.ats_screening import get_ats_screening_service
from app.services.db.supabase_service import get_db_admin_service
from app.services.email_queue import email_queue, Priority
from app.services.resume_parser import get_resume_parser
from app.utils.responses import api_error, ok

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/apply")


def _utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


@router.get("/job/{job_id}")
async def get_public_job(job_id: str):
    """Public job details. No auth required."""
    db = get_db_admin_service()

    jobs = await db.select(
        "job_descriptions",
        columns=(
            "id, title, role, level, description, must_have_skills, good_to_have_skills, "
            "min_experience_years, is_active, created_by"
        ),
        filters={"id": job_id, "is_active": True},
        limit=1,
    )
    if not jobs:
        return api_error(message="Job not found or no longer accepting applications", status_code=404)

    # Check candidate limit
    from app.utils.billing_helpers import check_candidate_limit
    err_msg = await check_candidate_limit(db, jobs[0].get("created_by"))
    if err_msg:
        return api_error(message="Candidates cannot be onboarded into this job", status_code=403)

    return ok(jobs[0])


@router.post("/submit")
async def submit_application(
    job_id: str = Form(...),
    full_name: str = Form(...),
    email: str = Form(...),
    phone: Optional[str] = Form(None),
    portfolio_url: Optional[str] = Form(None),
    github_url: Optional[str] = Form(None),
    location: Optional[str] = Form(None),
    vendorName: Optional[str] = Form(None),
    mainSkillset: Optional[str] = Form(None),
    consent_given: str = Form(...),
    resume: Optional[UploadFile] = File(None),
):
    """Submit a public job application. No auth required."""
    db = get_db_admin_service()

    if not job_id or not full_name or not email:
        return api_error(message="job_id, full_name, and email are required", status_code=400)

    consent = str(consent_given).lower() == "true"
    if not consent:
        return api_error(message="Consent is required to submit application", status_code=400)

    # Verify job exists and active (need full row for screening)
    job_rows = await db.select(
        "job_descriptions",
        columns="*",
        filters={"id": job_id, "is_active": True},
        limit=1,
    )
    if not job_rows:
        return api_error(message="Job not found or no longer accepting applications", status_code=404)
    job = job_rows[0]

    # Candidate lookup by email
    existing_candidates = await db.select(
        "candidates",
        columns="id",
        filters={"email": email},
        limit=1,
    )

    candidate_id: Optional[str] = None
    is_new_candidate = False
    is_new_for_recruiter = True
    recruiter_id = job.get("created_by")

    if existing_candidates:
        candidate_id = existing_candidates[0].get("id")

        # Check already applied
        existing_apps = await db.select(
            "job_applications",
            columns="id",
            filters={"candidate_id": candidate_id, "job_id": job_id},
            limit=1,
        )
        if existing_apps:
            return api_error(message="You have already applied to this job", status_code=400)
            
        def _check_recruiter_has_cand():
            return db.client.from_("job_applications").select("id, job_descriptions!inner(created_by)").eq("candidate_id", candidate_id).eq("job_descriptions.created_by", recruiter_id).limit(1).execute()
        cr_res = await db.run(_check_recruiter_has_cand)
        if getattr(cr_res, "data", None):
            is_new_for_recruiter = False
    else:
        is_new_candidate = True
        import uuid
        candidate_id = str(uuid.uuid4())

    if is_new_for_recruiter:
        from app.utils.billing_helpers import consume_candidate_slot
        err_msg = await consume_candidate_slot(db, recruiter_id, candidate_id, job_id)
        if err_msg:
            return api_error(message="Candidates cannot be onboarded into this job", status_code=403)

    try:

        # Parse resume (best-effort)
        resume_text = ""
        resume_parsed_data: Any = None
        if resume is not None:
            _resume_filename = resume.filename or "resume.pdf"
            try:
                content = await resume.read()
                parser = get_resume_parser()
                resume_text, parsed = await parser.parse_resume(content, _resume_filename)
                # Align with Node: store JSON-like dict, but keep any pydantic models serializable
                resume_parsed_data = parsed.model_dump() if hasattr(parsed, "model_dump") else parsed
                logger.info(
                    "apply: resume_parsed file=%s candidate_email=%s char_count=%d",
                    _resume_filename, email, len(resume_text),
                )
            except Exception as parse_exc:
                logger.error(
                    "apply: RESUME_PARSE_FAILED file=%s candidate_email=%s error=%s "
                    "– continuing without ATS screening",
                    _resume_filename, email, parse_exc,
                )
                resume_text = ""
                resume_parsed_data = None


        # Create or update candidate
        candidate_payload: Dict[str, Any] = {
            "full_name": full_name,
            "email": email,
            "phone": phone,
            "portfolio_url": portfolio_url,
            "github_url": github_url,
            "location": location,
            "vendorName": vendorName,
            "mainSkillset": mainSkillset,
            "consent_given": consent,
            "consent_timestamp": _utc_now_iso(),
            "updated_at": _utc_now_iso(),
        }
        if resume_text:
            candidate_payload["resume_text"] = resume_text
            candidate_payload["resume_parsed_data"] = resume_parsed_data

        if not is_new_candidate:
            # SECURITY: When updating an existing shared candidate record, only overwrite fields
            # that are currently null/empty. This prevents a new application from clobbering
            # contact details recorded by a different recruiter's tenant context.
            existing_full = await db.select("candidates", columns="*", filters={"id": candidate_id}, limit=1)
            existing_row = existing_full[0] if existing_full else {}

            safe_update: Dict[str, Any] = {"updated_at": _utc_now_iso()}
            # Always update consent (opt-in)
            if consent:
                safe_update["consent_given"] = True
                if not existing_row.get("consent_timestamp"):
                    safe_update["consent_timestamp"] = _utc_now_iso()
            # Only fill in missing contact fields
            for field, value in [
                ("full_name", full_name),
                ("phone", phone),
                ("portfolio_url", portfolio_url),
                ("github_url", github_url),
                ("location", location),
                ("vendorName", vendorName),
                ("mainSkillset", mainSkillset),
            ]:
                if value and not existing_row.get(field):
                    safe_update[field] = value
            # Resume data from this submission is always stored (application-specific)
            if resume_text:
                safe_update["resume_text"] = resume_text
                safe_update["resume_parsed_data"] = resume_parsed_data

            await db.update("candidates", safe_update, filters={"id": candidate_id})
        else:
            await db.insert(
                "candidates",
                [{"id": candidate_id, **candidate_payload, "created_at": _utc_now_iso()}],
            )

        # Create job application
        import uuid

        application_id = str(uuid.uuid4())
    
        app_payload = {
            "id": application_id,
            "candidate_id": candidate_id,
            "job_id": job_id,
            "status": "applied",
            "applied_at": _utc_now_iso(),
            "created_at": _utc_now_iso(),
        }

        # ISOLATION FIX: Always store the applicant's submitted identity fields in
        # candidate_overrides so their name/phone/etc. is scoped to this specific
        # job application rather than relying on the shared candidates row (which
        # is not overwritten when the email already exists).
        _overrides: Dict[str, Any] = {}
        if full_name:
            _overrides["full_name"] = full_name
        if phone:
            _overrides["phone"] = phone
        if location:
            _overrides["location"] = location
        if vendorName:
            _overrides["vendorName"] = vendorName
        if mainSkillset:
            _overrides["mainSkillset"] = mainSkillset
        if portfolio_url:
            _overrides["portfolio_url"] = portfolio_url
        if github_url:
            _overrides["github_url"] = github_url
        if _overrides:
            app_payload["candidate_overrides"] = _overrides

        await db.insert("job_applications", [app_payload])

        # ATS screening (best-effort) if resume parsed
        if resume_parsed_data:
            try:
                from app.models.schemas import JobDescription, ResumeData

                screening_service = get_ats_screening_service()

                resume_model = ResumeData(**resume_parsed_data) if isinstance(resume_parsed_data, dict) else ResumeData()
                job_model = JobDescription(
                    id=str(job.get("id")),
                    title=str(job.get("title")),
                    role=str(job.get("role") or "custom"),
                    level=job.get("level") or "mid",
                    description=str(job.get("description") or ""),
                    must_have_skills=job.get("must_have_skills") or [],
                    good_to_have_skills=job.get("good_to_have_skills") or [],
                    min_experience_years=int(job.get("min_experience_years") or 0),
                    resume_cutoff=int(job.get("resume_cutoff") or 0),
                    assessment_cutoff=int(job.get("assessment_cutoff") or 0),
                    interview_cutoff=int(job.get("interview_cutoff") or 0),
                    is_active=bool(job.get("is_active") is not False),
                    created_at=job.get("created_at"),
                    updated_at=job.get("updated_at"),
                )

                result = await screening_service.screen_candidate(resume_model, resume_text or "", job_model)

                screening_data = {
                    "candidate_id": candidate_id,
                    "job_id": job_id,
                    "overall_score": result.overall_score,
                    "skill_relevance_score": result.skill_relevance_score,
                    "experience_score": result.experience_score,
                    "education_score": result.education_score,
                    "credibility_score": result.credibility_score,
                    "shortlisted": result.shortlisted,
                    "shortlist_reason": result.shortlist_reason,
                    "reason_codes": [rc.model_dump() for rc in result.reason_codes],
                    "detailed_analysis": result.detailed_analysis.model_dump() if result.detailed_analysis else None,
                    "screened_at": _utc_now_iso(),
                    "updated_at": _utc_now_iso(),
                }

                existing_screenings = await db.select(
                    "ats_screenings",
                    columns="id",
                    filters={"candidate_id": candidate_id, "job_id": job_id},
                    limit=1,
                )
                if existing_screenings:
                    await db.update(
                        "ats_screenings",
                        screening_data,
                        filters={"id": existing_screenings[0].get("id")},
                    )
                else:
                    await db.insert(
                        "ats_screenings",
                        [{"id": str(uuid.uuid4()), **screening_data, "created_at": _utc_now_iso()}],
                    )
            except Exception:
                pass

        # Confirmation email (best-effort, async via queue)
        try:
            html, text, subject = email_queue.build_application_received(
                candidate_name=full_name,
                job_title=str(job.get("title") or ""),
            )
            await email_queue.enqueue(
                to_email=email,
                subject=subject,
                html_body=html,
                text_body=text,
                priority=Priority.NORMAL,
                idempotency_key=f"apply:{application_id}",
            )
        except Exception:
            pass

        return ok(
            {
                "id": application_id,
                "job_id": job_id,
                "candidate_id": candidate_id,
                "status": "applied",
                "message": f"Application submitted successfully for {str(job.get('title') or '')}. Check your email for confirmation.",
            },
            status_code=201,
        )

    except Exception as e:
        if is_new_for_recruiter:
            from app.utils.billing_helpers import refund_candidate_slot
            await refund_candidate_slot(db, recruiter_id)
        raise e
